cloudinit/osys/windows/network.py

```diff
@@ -0,0 +1,415 @@
+    for i in range(0, phys_address_len):
+        b = phys_address[i]
+        if mac_address:
+            mac_address += ":"
+        mac_address += "%02X" % b
+    return mac_address
+
+
+def _socket_addr_to_str(socket_addr):
+    addr_str_len = wintypes.DWORD(256)
+    addr_str = ctypes.create_unicode_buffer(256)
+
+    ret_val = ws2_32.WSAAddressToStringW(
+        socket_addr.lpSockaddr,
+        socket_addr.iSockaddrLength,
+        None, addr_str, ctypes.byref(addr_str_len))
+    if ret_val:
+        raise Exception("WSAAddressToStringW failed: %s"
+                        % ws2_32.WSAGetLastError())
+
+    return addr_str.value
+
+
+def _get_registry_dhcp_server(adapter_name):
+    with winreg.OpenKey(
+            winreg.HKEY_LOCAL_MACHINE,
+            "SYSTEM\\CurrentControlSet\\Services\\" +
+            "Tcpip\\Parameters\\Interfaces\\%s" % adapter_name, 0,
+            winreg.KEY_READ) as key:
+        try:
+            dhcp_server = winreg.QueryValueEx(key, "DhcpServer")[0]
+            if dhcp_server == "255.255.255.255":
+                dhcp_server = None
+            return dhcp_server
+        except Exception as ex:
+            # Not found
+            if ex.errno != 2:
+                raise
+
+
+class Network(network.BaseNetwork):
+    """Network namespace object tailored for the Windows platform."""
+
+    def routes(self):
+        """Get a collection of the available routes."""
+        return RouteCollection()
+
+    def default_gateway(self):
+        """Get the default gateway.
+
+        This will actually return a :class:`Route` instance. The gateway
+        can be accessed with the :attr:`gateway` attribute.
+        """
+        return next((r for r in self.routes() if r.destination == '0.0.0.0'),
+                    None)
+
+    def interfaces(self):
+        """Get a list of available interfaces."""
+
+        interfaces = []
+        size = wintypes.ULONG()
+        ret_val = iphlpapi.GetAdaptersAddresses(
+            ws2_32.AF_UNSPEC,
+            iphlpapi.GAA_FLAG_SKIP_ANYCAST,
+            None, None, ctypes.byref(size))
+
+        if ret_val == kernel32.ERROR_NO_DATA:
+            return interfaces
+
+        if ret_val == kernel32.ERROR_BUFFER_OVERFLOW:
+            proc_heap = kernel32.GetProcessHeap()
+            p = kernel32.HeapAlloc(proc_heap, 0, size.value)
+            if not p:
+                raise Exception("Cannot allocate memory")
+
+            ws2_32.init_wsa()
+
+            try:
+                p_addr = ctypes.cast(p, ctypes.POINTER(
+                    iphlpapi.IP_ADAPTER_ADDRESSES))
+
+                ret_val = iphlpapi.GetAdaptersAddresses(
+                    ws2_32.AF_UNSPEC,
+                    iphlpapi.GAA_FLAG_SKIP_ANYCAST,
+                    None, p_addr, ctypes.byref(size))
+
+                if ret_val == kernel32.ERROR_NO_DATA:
+                    return interfaces
+
+                if ret_val:
+                    raise Exception("GetAdaptersAddresses failed")
```

Comment:
Some other kind of exception would be useful. Perhaps should be a cloudinit/exceptions.py that has these exceptions. Just raising ``Exception`` doesn't  really allow anyone to catch specific errors/exceptions...